docs/developer/development/ci.rst

```diff
@@ -0,0 +1,146 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Continuous integration system is controlled by the ``eve/main.yml`` YAML file.
+
+A stage is defined by a worker and a list of steps. Each stage should be in
+the ``stages`` section and triggered by ``pre-merge`` or ``post-merge``.
+
+To know the different kind of workers available, all the builtin steps, how to
+trigger a stage, ... please refer to the eve documentation.
+
+A test stage in MetalK8s context
+""""""""""""""""""""""""""""""""
+
+In MetalK8s context each test stage (eve stage that represents a full test)
+should generate a status file containing the result of the test, either a
+success or a failure, and a JUnit file containing the result of the test
+and information about this test.
+
+To generate the JUnit file, each stage needs the following information:
+
+- The name of the Test Suite this test stage is part of.
+- Section path to regroup test in a Test Suite if needed (optional)
+- A test name
+
+Before executing all the steps of the test we first generate a failed
+result and at the end of the test we generate a success result. So that the
+failed result get overrided by the success one if everything goes well.
+
+At the very end, the final status of a test should be uploaded no matter the
+outcome of the test.
+
+To generate these result we already have several helpers available.
+
+Example:
+
+Consider we want a new test named ``My Test`` which is part of
+the subsection ``My sub section`` of the section ``My section`` in the
+test suite ``My Test Suite``.
+
+.. note::
+
+  Test, suite, class names are not case sensitive in ``eve/main.yml``.
+
+.. code-block:: yaml
+
+  my-stage:
+    _metalk8s_internal_info:
+      junit_info: &_my_stage_junit_info
+        TEST_SUITE: my test suite
+        CLASS_NAME: my section.my sub section
+        TEST_NAME: my test
+    worker:
+      # ...
+      # Worker informations
+      # ...
+    steps:
+      - Git: *git_pull
+      - ShellCommand: # Generate a failed final status
+          <<: *add_final_status_artifact_failed
+          env:
+            <<: *_env_final_status_artifact_failed
+            <<: *_my_stage_junit_info
+            STEP_NAME: my-stage
+       # ...
+       # All test steps should be here !
+       # ...
+       - ShellCommand: # Generate a success final status
+         <<: *add_final_status_artifact_success
+         env:
+           <<: *_env_final_status_artifact_success
+           <<: *_my_stage_junit_info
+           STEP_NAME: my-stage
+       - Upload: *upload_final_status_artifact
+
+TestRail upload
+"""""""""""""""
+
+To store result we use TestRail. TestRail is a declarative engine
+so before upload result you need to create all test Suites, Plans, Cases,
+Runs, ... you want to.
+
+.. warning::
+
+  TestRail upload is only done for Post-merge as we do not want to
+  store each and every test result coming from branch with some work not
+  completely finished, reviewed or approved.
```

Comment:
I agree with this point, but it makes me think: maybe we should also run the pre-merge tests during post-merge? So we still gather results and stats for these tests on merged code?

Reply:
I agree, we may want to trigger pre-merge during post-merge, but anyway we do not want to upload result during pre-merge, it's during post-merge always (just post-merge may trigger pre-merge)